[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data

n2h9 opened the pull request:
## Description
Contribution to this topic [Rich Disassembler for LLDB](https://discourse.llvm.org/t/rich-disassembler-for-lldb/76952), this part.
```
The rich disassembler output should be exposed as structured data and made available through LLDB’s scripting API so more tooling could be built on top of this
```

----

#### Update from 2025-11-27

I split this work into 3 parts:
1) https://github.com/llvm/llvm-project/pull/168276
2) https://github.com/llvm/llvm-project/pull/169408
3) https://github.com/llvm/llvm-project/pull/177676

And mark this pr as a draft to not to mislead reviewers, and continue with smaller ones;

----

<details>

<summary>Old description (before 2025-11-27)</summary>


## Current behaviour
Right now variable annotation is present during disassembling using lldb, this part `structured data and made available through LLDB’s scripting API` is not implemented yet. 

## This pr
- adds `annotateStructured` to `VariableAnnotator`;
- uses it in original `annotate` function;
- adds `GetVariableAnnotations(lldb::SBTarget target);` to `SBInstruction`;
- adds python  test to test functionality of `GetVariableAnnotations`

## Testing 
run test with 
```
./build/bin/lldb-dotest -v -t -p TestVariableAnnotationsDisassembler.py lldb/test/API/functionalities/disassembler-variables
```

all tests are passing, 9 existing and 1 newly added.
<details>
<summary>screenshot 2025-10-26</summary>
<img width="1235" height="742" alt="image" src="https://github.com/user-attachments/assets/829ff89a-e822-405d-adae-b6a19f7381cb" />
</details>

<details>
<summary>screenshot 2025-11-02</summary>
<img width="1396" height="782" alt="image" src="https://github.com/user-attachments/assets/36c69f42-9b82-4b2f-936c-405899aa081b" />

</details>


<details>
<summary>screenshot 2025-11-15</summary>
<img width="1306" height="377" alt="image" src="https://github.com/user-attachments/assets/3dc5005c-fecb-429b-bf6b-4540e958be67" />

</details>


</details>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data

Signed-off-by: Nikita B <n2h9z4@gmail.com>

JDevlieghere reviewed (changes requested):
It looks like this patch was created with the help of AI. There's nothing inherently wrong with that, except that there's a bunch of places where it is diverging from the prevailing coding style. I highlighted some of those issues in the comments, but please do another pass of the PR yourself and make sure the new code matches the surrounding code.

n2h9 committed:
Update lldb/include/lldb/Core/Disassembler.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

n2h9 committed:
Update lldb/include/lldb/Core/Disassembler.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

n2h9 committed:
Update lldb/test/API/functionalities/disassembler-variables/TestVariableAnnotationsDisassembler.py

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: remove braces around single line ifs and cycles

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: wrap print instructions with if self.TraceOn():

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: end comments with dots in python file

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: rename class method to much declaration and LLDB's coding style

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
Merge branch 'main' into rich-disassembler-structured-variable-annotations-00

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: update comments to be Doxygen-style

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: remove VarState and keep only VariableAnnotation struct

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: update both annotate and AnnotateStructured to use shared pointer tp module by value

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: update both annotate to Annotate class function name to follow lldb coding style

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: update descriptive comment for AnnotateStructured

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: use AddressRange instead of start_address and end_address

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 commented on lldb/include/lldb/Core/Disassembler.h:
```diff
@@ -566,6 +566,21 @@ class Disassembler : public std::enable_shared_from_this<Disassembler>,
+      register_kind; // Register numbering scheme for location interpretation
+  std::optional<std::string>
+      decl_file;                     // Source file where variable was declared
+  std::optional<uint32_t> decl_line; // Line number where variable was declared
```
The idea is to provide an information in this fields on  where the variable is declared, like we  we have in a DWARF output.
The data comes from [Declaration](https://github.com/llvm/llvm-project/blob/main/lldb/include/lldb/Core/Declaration.h#L124) object which returns line number as uint.

We only have (and probably only need for the purpose of variable annotation) information on file name, line number and column number from Declaration object. While LineEntry suppose to contain more information in it. 

I think it’s better to keep this as a uint. What do you think?

n2h9 commented on lldb/include/lldb/Core/Disassembler.h:
```diff
@@ -566,6 +566,21 @@ class Disassembler : public std::enable_shared_from_this<Disassembler>,
+  std::string variable_name;
+  std::string location_description; // e.g., "r15", "undef", "const_0"
+  lldb::addr_t start_address;       // Where this annotation starts being valid
+  lldb::addr_t end_address;         // Where this annotation ends being valid
```
Valid point, updated :white_check_mark: , thank you :bowing_woman:

n2h9 commented on lldb/include/lldb/Core/Disassembler.h:
```diff
@@ -574,16 +589,39 @@ class VariableAnnotator {
+
+    std::optional<std::string> decl_file;
+    std::optional<uint32_t> decl_line;
+    std::optional<std::string> type_name;
```
Yes,  I agree. The idea was to have different structs for internal and external usage to have flexibility to have different structure of data. But probably this not necessary at all. 

Updated ✅ , thank you 🙇‍♀️.

n2h9 commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -302,14 +302,38 @@ bool Disassembler::ElideMixedSourceAndDisassemblyLine(
+}
+
+std::vector<VariableAnnotation>
+VariableAnnotator::annotateStructured(Instruction &inst, Target &target,
```
Thanks Copilot, updated :white_check_mark: .

Also updated existing `annotate` method to be `Annotate` to follow the `LLDB's UpperCamelCase naming convention for methods` :hugs: .

n2h9 commented:
Hey hey 

> > It looks like this patch was created with the help of AI. There's nothing inherently wrong with that, except that there's a bunch of places where it is diverging from the prevailing coding style. I highlighted some of those issues in the comments, but please do another pass of the PR yourself and make sure the new code matches the surrounding code.
> 
> Thanks for your review @JDevlieghere 🤗 !
> 
> Let me address the comments and I will double check to be sure the code is matching prevailing coding style 😇 .

Hey hey @JDevlieghere hello.
I have addressed the comments :blush: could you please take a look once  you will have some time :hugs: .

Thank you :bowing_woman:

JDevlieghere commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -299,17 +299,45 @@ bool Disassembler::ElideMixedSourceAndDisassemblyLine(
+                ? llvm::formatv("<{0}>", VariableAnnotator::kUndefLocation)
+                      .str()
+                : annotation.location_description)
+            .str();
```
This seems like it could benefit from writing this as a stream:
```
  std::string display_string;
  llvm::raw_string_ostream os(display_string);
  os << annotation.variable_name;
  [...]
  events.push_back(std::move(display_string));
```

n2h9 committed:
Update lldb/include/lldb/Core/Disassembler.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: rename Live_ to m_live_vars to follow lldb codding style

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: rename Current to current_vars to follow lldb codding style and to be consistent with m_live_vars

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: replace llvm::formatv with stream, formatting

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: simplify test_structured_annotations_api

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: remove target and module from Annotate methods signature

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 commented:
While updating the test I have noticed that we receive structured annotations by calling `annotations = inst.GetVariableAnnotations(target)`. 
Which is probably slightly redundant, because we receive instructions from target, and then variable annotations from particular instructions, but still need to pass target :thinking: .

The reason for this is that the original  [annotate](https://github.com/llvm/llvm-project/blob/main/lldb/include/lldb/Core/Disassembler.h#L589) method expects a target and module. 

I think we can simplify and keep only `instruction` as method param, and get `module` from address inside the method, I updated in this [commit](https://github.com/llvm/llvm-project/pull/165163/commits/651b407a5f955b21b9b514811570cb3f07df7b3e) let me know if it makes sense :innocent: .

Thank you :bowing_woman: .

n2h9 committed:
Merge branch 'main' into rich-disassembler-structured-variable-annotations-00

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: update TestVariableAnnotationsDisassembler.py formatting with darker

Signed-off-by: Nikita B <n2h9z4@gmail.com>

n2h9 committed:
[lldb] [disassembler] chore: enhance VariableAnnotator to return structured data: merge main, fix conflicts

Signed-off-by: Nikita B <n2h9z4@gmail.com>
